### ISSUE TITLE
Example for special edge case and some comments on the paper

### ISSUE THREAD

**Issue by davidot** [fastfloat/fast_float#146]
While implementing the algorithm as described in the paper I found a couple of things I didn't understand and/or which are mismatched with the code in this repository.

- In step 1 and 2 of the algorithm only the 64 bit case is covered.
- After step 13 we could have rounded a denormal up to a normal and this case is not explicitly mentioned here, I do vaguely remember it somewhere else so maybe this was intentional.
- In step 20 of the algorithm it says: `if m = 254 (64-bit) or m = 225 (32-bit) then m = m ÷ 2; p ← p + 1 end i`, but both the [code in the repo](https://github.com/fastfloat/fast_float/blob/main/include/fast_float/decimal_to_binary.h#L179) and running the tests mean using this is incorrect. It seems to be `m >= 2^53`, and since only the 54th bit is cleared after any overflow would actually stay in the mantissa.

Finally I have now parsed all of the supplemental files and tried a bunch of random value but have never hit the special case in step 6. `if z mod 264 = 264 − 1 and q < [−27, 55] then Abort end if`.
Do you happen to have cases which hit this? Ideally one within the safe exponent and without outside, (although those are probably just the same case with a different exponent)

**Comment by lemire** [comment]
> have never hit the special case in step 6

That's correct. This case is never encountered. We have a submitted manuscript with a proof.

**Comment by lemire** [comment]
> In step 1 and 2 of the algorithm only the 64 bit case is covered.

In what sense ?

**Comment by davidot** [comment]
> > have never hit the special case in step 6
> 
> That's correct. This case is never encountered. We have a submitted manuscript with a proof.

Good to know! So the fall back in the code is just an illustration how you would recover there?

> > In step 1 and 2 of the algorithm only the 64 bit case is covered.
> 
> In what sense ? Can you illustrate how you get a wrong result with an example?

You will never get a wrong example sorry, just that since finite 32 bit floats only goes from 10e--45 up to 10e38, so for 32 bit these constant can be a lot smaller (like you already have in the code)

> > After step 13 we could have rounded a denormal up to a normal
> 
> Can you illustrate this issue with an example?

I may have actually been mistaken about this point, sorry. I think I was just doing something incorrectly.

**Comment by lemire** [comment]
> In step 20 of the algorithm it says

You are correct, it is a typo in the original paper.

**Comment by lemire** [comment]
> Good to know! So the fall back in the code is just an illustration how you would recover there?

The fallback is a legacy of a time when I was unsure whether it was needed. Before removing it, I wanted to have a formal proof. We now have such a proof and we can remove the fallback.

It has not been removed from the implementations.

**Comment by lemire** [comment]
I am closing this issue. Thanks for your comments. If you have more comments, just open a new issue.

**Comment by lemire** [comment]
(You can also use the 'discussion' feature on GitHub.)
